In [15]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [16]:
import warnings
warnings.filterwarnings('ignore')

In [17]:
df = pd.read_csv(r"C:\Users\HP\Desktop\dataset\13_dropout_layer\sonar_dataset.csv", header=None)
df.sample(5)

0       1       2       3       4       5       6       7       8   \
95   0.0291  0.0400  0.0771  0.0809  0.0521  0.1051  0.0145  0.0674  0.1294   
20   0.0473  0.0509  0.0819  0.1252  0.1783  0.3070  0.3008  0.2362  0.3830   
92   0.0260  0.0192  0.0254  0.0061  0.0352  0.0701  0.1263  0.1080  0.1523   
165  0.0221  0.0065  0.0164  0.0487  0.0519  0.0849  0.0812  0.1833  0.2228   
67   0.0368  0.0403  0.0317  0.0293  0.0820  0.1342  0.1161  0.0663  0.0155   

         9   ...      51      52      53      54      55      56      57  \
95   0.1146  ...  0.0056  0.0237  0.0204  0.0050  0.0137  0.0164  0.0081   
20   0.3759  ...  0.0193  0.0118  0.0064  0.0042  0.0054  0.0049  0.0082   
92   0.1630  ...  0.0118  0.0120  0.0051  0.0070  0.0015  0.0035  0.0008   
165  0.1810  ...  0.0089  0.0051  0.0015  0.0075  0.0058  0.0016  0.0070   
67   0.0506  ...  0.0091  0.0160  0.0160  0.0081  0.0070  0.0135  0.0067   

         58      59  60  
95   0.0139  0.0111   R  
20   0.0028  0.0027   R  
92   0.0044  0.0077   R  
165  0.0074  0.0038   M  
67   0.0078  0.0068   R  

[5 rows x 61 columns]

In [18]:
df.shape

(208, 61)

In [19]:
df.isnull().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [20]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [31]:
y = pd.get_dummies(y, drop_first=True)
y.head(10) # R --> 1 and M --> 0

1
0  1
1  1
2  1
3  1
4  1
5  1
6  1
7  1
8  1
9  1

In [32]:
y.value_counts()

1
0    111
1     97
dtype: int64

In [33]:
y

1
0    1
1    1
2    1
3    1
4    1
..  ..
203  0
204  0
205  0
206  0
207  0

[208 rows x 1 columns]

In [34]:
X.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      50      51      52      53      54      55      56  \
0  0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1  0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2  0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3  0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4  0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   

       57      58      59  
0  0.0084  0.0090  0.0032  
1  0.0049  0.0052  0.0044  
2  0.0164  0.0095  0.0078  
3  0.0044  0.0040  0.0117  
4  0.0048  0.0107  0.0094  

[5 rows x 60 columns]

In [35]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [36]:
X_train.head()

0       1       2       3       4       5       6       7       8   \
67   0.0368  0.0403  0.0317  0.0293  0.0820  0.1342  0.1161  0.0663  0.0155   
14   0.0124  0.0433  0.0604  0.0449  0.0597  0.0355  0.0531  0.0343  0.1052   
164  0.0163  0.0198  0.0202  0.0386  0.0752  0.1444  0.1487  0.1484  0.2442   
179  0.0394  0.0420  0.0446  0.0551  0.0597  0.1416  0.0956  0.0802  0.1618   
19   0.0126  0.0149  0.0641  0.1732  0.2565  0.2559  0.2947  0.4110  0.4983   

         9   ...      50      51      52      53      54      55      56  \
67   0.0506  ...  0.0058  0.0091  0.0160  0.0160  0.0081  0.0070  0.0135   
14   0.2120  ...  0.0078  0.0083  0.0057  0.0174  0.0188  0.0054  0.0114   
164  0.2822  ...  0.0027  0.0077  0.0026  0.0031  0.0083  0.0020  0.0084   
179  0.2558  ...  0.0118  0.0146  0.0040  0.0114  0.0032  0.0062  0.0101   
19   0.5920  ...  0.0153  0.0092  0.0035  0.0098  0.0121  0.0006  0.0181   

         57      58      59  
67   0.0067  0.0078  0.0068  
14   0.0196  0.0147  0.0062  
164  0.0108  0.0083  0.0033  
179  0.0068  0.0053  0.0087  
19   0.0094  0.0116  0.0063  

[5 rows x 60 columns]

In [39]:
import tensorflow as tf
from tensorflow import keras

In [40]:

model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 786us/step - loss: 0.6881 - accuracy: 0.5004
Epoch 2/100
20/20 [==============================] - 0s 787us/step - loss: 0.6641 - accuracy: 0.5474
Epoch 3/100
20/20 [==============================] - 0s 682us/step - loss: 0.6382 - accuracy: 0.6297
Epoch 4/100
20/20 [==============================] - 0s 735us/step - loss: 0.6010 - accuracy: 0.7994
Epoch 5/100
20/20 [==============================] - 0s 682us/step - loss: 0.5672 - accuracy: 0.7849
Epoch 6/100
20/20 [==============================] - 0s 735us/step - loss: 0.5530 - accuracy: 0.7814
Epoch 7/100
20/20 [==============================] - 0s 682us/step - loss: 0.5290 - accuracy: 0.7850
Epoch 8/100
20/20 [==============================] - 0s 682us/step - loss: 0.4784 - accuracy: 0.7777
Epoch 9/100
20/20 [==============================] - 0s 682us/step - loss: 0.5155 - accuracy: 0.7656
Epoch 10/100
20/20 [==============================] - 0s 735us/step - loss: 0.4495 - accura

20/20 [==============================] - 0s 735us/step - loss: 0.0116 - accuracy: 1.0000
Epoch 82/100
20/20 [==============================] - 0s 735us/step - loss: 0.0210 - accuracy: 1.0000
Epoch 83/100
20/20 [==============================] - 0s 682us/step - loss: 0.0093 - accuracy: 1.0000
Epoch 84/100
20/20 [==============================] - 0s 577us/step - loss: 0.0093 - accuracy: 1.0000
Epoch 85/100
20/20 [==============================] - 0s 630us/step - loss: 0.0055 - accuracy: 1.0000
Epoch 86/100
20/20 [==============================] - 0s 682us/step - loss: 0.0061 - accuracy: 1.0000
Epoch 87/100
20/20 [==============================] - 0s 683us/step - loss: 0.0066 - accuracy: 1.0000
Epoch 88/100
20/20 [==============================] - 0s 682us/step - loss: 0.0060 - accuracy: 1.0000
Epoch 89/100
20/20 [==============================] - 0s 682us/step - loss: 0.0042 - accuracy: 1.0000
Epoch 90/100
20/20 [==============================] - 0s 682us/step - loss: 0.0064 - accuracy: 

In [41]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 2ms/step - loss: 1.3771 - accuracy: 0.7308


[1.377098560333252, 0.7307692170143127]

In [42]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[1.4156804e-08 9.9909860e-01 9.8713028e-01 2.0703787e-06 9.9999940e-01
 9.9991369e-01 8.9063048e-02 9.9999952e-01 3.2211688e-06 9.9999964e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [46]:

y_test[:10]

1
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [47]:

from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.70      0.85      0.77        27
           1       0.79      0.60      0.68        25

    accuracy                           0.73        52
   macro avg       0.74      0.73      0.72        52
weighted avg       0.74      0.73      0.73        52



# Model with Dropout Layer

In [51]:

modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 0s 821us/step - loss: 0.6951 - accuracy: 0.4960
Epoch 2/100
20/20 [==============================] - 0s 735us/step - loss: 0.7161 - accuracy: 0.4478
Epoch 3/100
20/20 [==============================] - 0s 735us/step - loss: 0.6733 - accuracy: 0.5430
Epoch 4/100
20/20 [==============================] - 0s 735us/step - loss: 0.6931 - accuracy: 0.4720
Epoch 5/100
20/20 [==============================] - 0s 787us/step - loss: 0.6887 - accuracy: 0.5216
Epoch 6/100
20/20 [==============================] - 0s 787us/step - loss: 0.6728 - accuracy: 0.5128
Epoch 7/100
20/20 [==============================] - 0s 800us/step - loss: 0.6858 - accuracy: 0.5340
Epoch 8/100
20/20 [==============================] - 0s 840us/step - loss: 0.6701 - accuracy: 0.5796
Epoch 9/100
20/20 [==============================] - 0s 787us/step - loss: 0.6863 - accuracy: 0.4862
Epoch 10/100
20/20 [==============================] - 0s 787us/step - loss: 0.6821 - accura

20/20 [==============================] - 0s 735us/step - loss: 0.2977 - accuracy: 0.9052
Epoch 82/100
20/20 [==============================] - 0s 735us/step - loss: 0.2998 - accuracy: 0.9048
Epoch 83/100
20/20 [==============================] - 0s 735us/step - loss: 0.3032 - accuracy: 0.8876
Epoch 84/100
20/20 [==============================] - 0s 663us/step - loss: 0.2721 - accuracy: 0.9220
Epoch 85/100
20/20 [==============================] - 0s 630us/step - loss: 0.2663 - accuracy: 0.9310
Epoch 86/100
20/20 [==============================] - 0s 682us/step - loss: 0.2826 - accuracy: 0.9236
Epoch 87/100
20/20 [==============================] - 0s 682us/step - loss: 0.2577 - accuracy: 0.9029
Epoch 88/100
20/20 [==============================] - 0s 682us/step - loss: 0.3184 - accuracy: 0.8892
Epoch 89/100
20/20 [==============================] - 0s 682us/step - loss: 0.2987 - accuracy: 0.8725
Epoch 90/100
20/20 [==============================] - 0s 716us/step - loss: 0.2572 - accuracy: 

In [52]:
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 999us/step - loss: 0.4379 - accuracy: 0.7885


[0.43793752789497375, 0.7884615659713745]

Training Accuracy is still good but Test Accuracy Improved

In [53]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[3.3271255e-05 7.3523188e-01 8.6828434e-01 1.1776060e-02 9.9959660e-01
 9.2186004e-01 2.5493187e-01 9.9967623e-01 5.5779517e-03 9.9957108e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [54]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.74      0.93      0.82        27
           1       0.89      0.64      0.74        25

    accuracy                           0.79        52
   macro avg       0.81      0.78      0.78        52
weighted avg       0.81      0.79      0.78        52



You can see that by using dropout layer test accuracy increased from 0.73 to 0.79